src/main/java/frc/robot/hardware/signal/InputSignal.java

```diff
@@ -0,0 +1,70 @@
+package frc.robot.hardware.signal;
+
+import edu.wpi.first.math.geometry.Rotation2d;
+import org.littletonrobotics.junction.LogTable;
+import org.littletonrobotics.junction.inputs.LoggableInputs;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.function.DoubleFunction;
+
+
+public abstract class InputSignal<T> implements LoggableInputs {
+
```

Comment:
Before laying out instances of how the class functions, it makes much more sense to describe _what_ the class do and how it functions. Hence, it makes more sense to put the abstract class at the end of the class, or in a  different file.